Allow admin to set organization language at creation time

HemantUmre12 opened the pull request:
Commit-1: Adds a new dropdown, 'Organization language,' to the `/new` and `/realm/register` pages and utilizes it to set the `default_language` field of the `Realm` model.

Commit-2: Uses the `Realm.default_language` value, which is set by selecting the 'Organization language,' to internationalize the introductory messages of the initial streams.

Fixes: #25729.

<!-- Describe your pull request here.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Screenshots and screen captures</summary>



https://github.com/zulip/zulip/assets/87542880/f989f46d-2055-4fe3-90d5-860b476cb233



</details>


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

HemantUmre12 commented on zerver/context_processors.py:
```diff
@@ -259,8 +261,19 @@ def latest_info_context() -> Dict[str, str]:
+    # Normalize the language names in the language list
+    language_list = get_language_list()
+    for language in language_list:
+        language["name"] = unicodedata.normalize("NFC", language["name"])
```
I added this since I received the following error. I don't know if it's the correct way to fix it.
````
DROP DATABASE
nu.validator.servlet.VerifierServletTransaction - Starting static initializer.
nu.validator.servlet.VerifierServletTransaction - Reading miscellaneous properties.
nu.validator.servlet.VerifierServletTransaction - Starting to loop over config file lines.
nu.validator.servlet.VerifierServletTransaction - Finished reading config.
nu.validator.servlet.VerifierServletTransaction - Converted config to arrays.
nu.validator.servlet.VerifierServletTransaction - Prepared namespace array.
nu.validator.servlet.VerifierServletTransaction - Parsed doctype numbers into ints.
nu.validator.servlet.VerifierServletTransaction - The cache path prefix is: null
nu.validator.servlet.VerifierServletTransaction - Parsing set up. Starting to read schemas.
nu.validator.servlet.VerifierServletTransaction - Will load schema: http://s.validator.nu/html5.rnc
CREATE DATABASE
nu.validator.servlet.VerifierServletTransaction - Will load schema: http://s.validator.nu/html5-all.rnc
nu.validator.servlet.VerifierServletTransaction - Will load schema: http://s.validator.nu/xhtml5.rnc

Waiting for test server (may take a while).nu.validator.servlet.VerifierServletTransaction - Will load schema: http://s.validator.nu/xhtml5-all.rnc
nu.validator.servlet.VerifierServletTransaction - Will load schema: http://s.validator.nu/svg-xhtml5-rdf-mathml.rnc
.nu.validator.servlet.VerifierServletTransaction - Schemas read.
nu.validator.servlet.VerifierServletTransaction - Reading spec.
..........nu.validator.servlet.VerifierServletTransaction - Spec read.
.
WARNING: Future checker releases will bind by default to 127.0.0.1.
Your checker deployment might become unreachable unless you use the
nu.validator.servlet.bind-address system property or --bind-address
script option to bind the checker to a different address:

  python ./checker.py --bind-address 172.18.0.2 run
  java -Dnu.validator.servlet.bind-address=172.18.0.2 -cp vnu.jar nu.validator.servlet.Main 8888
  vnu-runtime-image/bin/java -Dnu.validator.servlet.bind-address=172.18.0.2 nu.validator.servlet.Main 8888
  vnu-runtime-image\bin\java.exe -Dnu.validator.servlet.bind-address=172.18.0.2 nu.validator.servlet.Main 8888

Checker service started at http://127.0.0.1:9988/


--- SERVER IS UP! ---

2023-06-25 16:01:04 [portico_documentation_crawler] ERROR: "http://localhost:9981/new/":243.46-243.57: info: Text run is not in Unicode Normalization Form C.

Failed

Error: Process completed with exit code 1.
````

sahil839 replied:
It is not clear to me when you are seeing this error. Are you seeing this on locally or on CI?

HemantUmre12 replied:
Having passed all tests locally, I encounter this error during CI.

sahil839 replied:
I am not sure as well that why the test is failing. You can ask in #automated-testing stream on CZO. Sharing a link to the failed job might be helpful. I can see this at https://github.com/zulip/zulip/actions/runs/5370569466/jobs/9742740276#step:16:98 but I am not sure about who all can access this.

timabbott replied:
So adding this bit of code to see which language has unicode normalization inconsistency, I get this:

`Tiếng Việt Tiếng Việt False`

```
                normalized = unicodedata.normalize("NFC", lang_info["name_local"])                      
                print(lang_info["name_local"], normalized, normalized == lang_info["name_local"])       
```

andersk replied:
Linking a previous discussion: [#automated testing > Unicode normalization error](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/Unicode.20normalization.20error/near/1600910).

timabbott replied:
(In the future, @HemantUmre12, please cross-link in both directions when you start a conversation about some detail in a PR)

HemantUmre12 commented:
@Ujjawal3, could you please take a look at this PR?

@zulipbot add "buddy review"

Ujjawal3 commented:
@HemantUmre12 I did review of some files. Till now the changes looks good to me. I will complete the review by today evening. Can you tell how to do the manual testing of the changes? I tried to test but there is no option for choosing default language while creating new realm in dev environment.

HemantUmre12 commented:
@Ujjawal3, you can use `./manage.py generate_realm_creation_link` to simulate the onboarding process.

sahil839 commented:
@Ujjawal3 you can also use the "New organization" link which is available in the footer when you log out in the dev server.

Ujjawal3 commented:
@HemantUmre12 I did the manual testing and reviewed the code changes. Almost everything looks good to me. Just a few minor points:
- As mentioned in issue description add a `?` button linking to `/help/configure-organization-language`.
- I think it will be better to mention in second commit message that language for introductory message will be the default language chosen while creating realm.

sahil839 commented on zerver/models.py:
```diff
@@ -2342,6 +2342,10 @@ class PreregistrationRealm(models.Model):
         default=Realm.ORG_TYPES["unspecified"]["id"],
         choices=[(t["id"], t["name"]) for t in Realm.ORG_TYPES.values()],
     )
+    org_language = models.CharField(
```
We should use `default_language` here so that we have the same field name in `PreregistrationRealm` and `Realm` objects.

sahil839 commented on zerver/views/registration.py:
```diff
@@ -839,6 +857,7 @@ def new_realm_send_confirm(
     email: str = REQ("email"),
     realm_name: str = REQ(str_validator=check_capped_string(Realm.MAX_REALM_NAME_LENGTH)),
     realm_type: int = REQ(json_validator=check_int_in(Realm.ORG_TYPE_IDS)),
+    realm_language: str = REQ(str_validator=check_capped_string(MAX_LANGUAGE_ID_LENGTH)),
```
`realm_default_language` should be more clear than `realm_language`.

sahil839 reviewed (commented):
Did an initial review of this. Mostly all comments are related to variable naming. Basically variables related to `Realm` and `PreregistrationRealm` obejcts should be named as `default_lanugage` and the ones related to form and templates can be named as `realm_default_language` or probably `default_language` itself. You should change the function parameters and elements ID, etc. as well accordingly.

Also, one thing I noticed is that the dropdown in realm creation form has more options than we have in the `Organization settings` panel dropdown. Both should have same options.

I will try to look at the CI error by looking at previous logs.

HemantUmre12 commented:
@sahil839, this PR is ready for another round of review. The changes since your last review are listed below:

- According to the discussion [here](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/Unicode.20normalization.20error/near/1600910), I have added a separate commit for Unicode normalization of language names when creating `language_name_map.json`. I think normalizing the rest of the message when pulling from transfix can be tackled in a separate PR, as we do not get language names particularly from transfix.

> Also, one thing I noticed is that the dropdown in realm creation form has more options than we have in the Organization settings panel dropdown.

- There we are allowing languages with a `translation percentage >= 5%` as an option. So, I have created `get_filtered_language_list()` to do the same.

- Renamed the variables as you suggested.

- Added a test for translation in the last commit.

sahil839 commented on zerver/tests/test_signup.py:
```diff
@@ -1752,6 +1752,17 @@ def test_create_realm_with_custom_language(self) -> None:
 
+        # Check welcome messages
+        for stream_name, text, message_count in [
+            (Realm.DEFAULT_NOTIFICATION_STREAM_NAME, "con l'argomento", 3),
```
Would be better to this text in a variable for readability like `italian_translation_for_private_stream` or something similar.

sahil839 commented on zerver/forms.py:
```diff
@@ -131,13 +132,17 @@ class RealmDetailsForm(forms.Form):
         del kwargs["realm_creation"]
 
         super().__init__(*args, **kwargs)
+        self.fields["realm_default_language"] = forms.ChoiceField(
```
We use a simple `CharField` for `default_languge` in models.py, so that would have worked fine too. Let's keep it like this for now and update after some more feedback.

sahil839 commented on templates/zerver/realm_creation_form.html:
```diff
@@ -34,6 +34,29 @@
+                {% else %}
+                    {% for language in language_list %}
+                        <option value="{{ language.code }}" {% if default_language_code == language.code %}selected{% endif %} >{{ _(language.name) }}</option>
+                    {% endfor %}
```
Would have been good to avoid this duplication here. Can we just pass the default code in `RealmCreationForm` as `initial_data`?

sahil839 reviewed (commented):
Posted a couple of comments. Looks in a good state overall and we should be able to finish it soon.
I am not sure about the normalization change commit though.

HemantUmre12 commented:
@sahil839 made the changes. Ready for another round of review.

sahil839 reviewed (commented):
Posted one comment regarding tests. Looks good otherwise.

sahil839 commented:
I noticed that the user's language would not be the same as that selected for the realm if the user selected a language other than the browser default for realm. I am not sure but I think this might confuse users. If we set the user's language also to the realm language, it will not be consistent with the other users who sign up after the realm is created, as for them we give preference to the browser language and then the realm default if needed.

sahil839 commented:
Looks good to me.

Now reading the code again I am not sure how to handle the invalid language same. By reading your test, I was considering the case where invalid language can be passed when creating realm through API, but on reading the code again I am not sure that URL can be used as an API endpoint by clients, since the url returns template or a redirect as a response.

And since we use a dropdown in the UI, I think we can just remove the error message code from template and also probably the test for that case. I also checked the existing code for handling organization type and we do not test the invalid type case. We can also wait for Tim's feedback if you want or you can just remove the test and the error message code in template. Sorry for the inconvenience.

HemantUmre12 commented:
@sahil839, I have removed the error code for now to maintain consistency. We can add that separately along with the organization type if we want.

sahil839 commented:
Other than the above point, the PR looks good to me. @timabbott I think this PR is ready for you to have a look and would be good to have your feedback on a couple of points discussed above.

timabbott commented:
Sorry for the long delay in integrating this! I rebased the migrations and other small tweaks, and added a couple commits on top designed to fix a weird focus experience problem noted while testing it. 

Thanks for doing this @HemantUmre12!

timabbott commented:
I removed the `get_filtered_language_list` function, since the 5% filtering should be being done in what languages we commit to the codebase (we sync using `tx pull -a -f --mode=translator --minimum-perc=5`), and thus `get_language_list` already has that filtered.

https://github.com/zulip/zulip/pull/26102#discussion_r1242031069
